### PR TITLE
ringbuffer + gitignore

### PR DESCRIPTION
**Pull request by dyldonahue**

### CODE REVIEW

**Review by nwdepatie — CHANGES_REQUESTED** [review]
Some stuff to address

File: middleware/include/ringbuffer.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+ringbuffer_t* ringbuffer_create(size_t capacity, size_t element_size);
+
+int ringbuffer_is_empty(ringbuffer_t* rb);
+
+int ringbuffer_is_full(ringbuffer_t* rb);
+
+int ringbuffer_enqueue(ringbuffer_t* rb, void* data);
+
+void* ringbuffer_dequeue(ringbuffer_t* rb, void* data);
```
Comment (nwdepatie) [inline middleware/include/ringbuffer.h:27]: Add in ringbuffer_destroy and keep naming consistent

**Review by nwdepatie — CHANGES_REQUESTED** [review]
Some more stuff

File: middleware/include/ringbuffer.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#ifndef RINGBUFFER_H
+#define RINGBUFFER_H
+
+#include <stddef.h>
+#include <stdint.h>
+
+#endif // RINGBUFFER_H
```
[the remark this answers is not in the document]
Comment (dyldonahue) [reply middleware/include/ringbuffer.h:7]: lmao we dont talk about that

File: middleware/src/ringbuffer.c [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ringbuffer_t* ringbuffer_create(size_t capacity, size_t element_size)
+{
+    ringbuffer_t* rb = (ringbuffer_t*)calloc(sizeof(rb));
+    if (rb == NULL) {
+
+        // Handle memory allocation failure
+        return NULL;
+    }
+
+    rb->buffer = (void**)malloc(capacity * sizeof(void*));
```
[the remark this answers is not in the document]
Comment (dyldonahue) [reply middleware/src/ringbuffer.c:12]: whoops this was a remnant of when i was doinf all at once good catch

File: middleware/src/ringbuffer.c [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    rb->element_size = element_size;
+
+    return rb;
+}
+
+void destroy_ringbuffer(ringbuffer_t* rb) 
+{
+    // Free each dynamically allocated element
+    for (size_t i = 0; i < rb->count; ++i) {
+        free(rb->buffer[(rb->front + i) % rb->capacity]);
```
[the remark this answers is not in the document]
Comment (dyldonahue) [reply middleware/src/ringbuffer.c:28]: somehwat redundant since the count param should handle it but no harm to adding

**Review by nwdepatie — APPROVED** [review]
I think this looks good now, lets merge in to get shepherd shmovin

**Pull request merged**